lib/puppet/util/selinux.rb

```diff
@@ -68,11 +69,36 @@ def get_selinux_default_context(file, resource_ensure = nil)
     end
 
     retval = Selinux.matchpathcon(file, mode)
-    if retval == -1
-      return nil
+    retval == -1 ? nil : retval[1]
+  end
+
+  def get_selinux_default_context_with_handle(file, handle, resource_ensure = nil)
+    return nil unless selinux_support?
+    # If the filesystem has no support for SELinux labels, return a default of nil
+    # instead of what selabel_lookup would return
+    return nil unless selinux_label_support?(file)
+    # If the file exists we should pass the mode to selabel_lookup for the most specific
+
+    # Handle is needed for selabel_lookup
+    raise ArgumentError, _("Cannot get default context with nil handle") unless handle
+
+    # If the file exists we should pass the mode to selabel_lookup for the most specific
+    # matching.  If not, we can pass a mode of 0.
+    begin
+      filestat = file_lstat(file)
+      mode = filestat.mode
+    rescue Errno::EACCES
+      mode = 0
+    rescue Errno::ENOENT
+      if resource_ensure
+        mode = get_create_mode(resource_ensure)
+      else
+        mode = 0
+      end
     end
 
-    retval[1]
+    retval = Selinux.selabel_lookup(handle, file, mode)
```

Comment:
The `matchpathcon` method accepts a `mode` argument https://manpages.ubuntu.com/manpages/jammy/en/man3/matchpathcon.3.html

But `selabel_lookup` accepts different arguments:
https://manpages.ubuntu.com/manpages/jammy/en/man3/selabel_lookup.3.html

I assume `key` is the file path, but am not sure about `type`. Also the C API has both an integer return value and a `security_context_t *context` output parameter. So we need to understand how that gets passed back through the ruby bindings.

Also it says: "returning the result in the memory pointed to by context, which must be freed by the caller using [freecon(3)](https://www.unix.com/man-page/debian/3/freecon/)." So we need to check if swig generates code to call `freecon` for us or if we need to do it

Reply:
The C `selabel_lookup` API takes 4 arguments

```
       int selabel_lookup(
                          struct selabel_handle *hnd,
                          char **context,
                          const char *key,
                          int type);
```

The `hnd` is whatever handle `selabel_open` returned.
The `context` is an output parameter that receives the the security context (string)
The `key` is the file path
The `type` should be 0 as best I can tell from looking at how `/usr/bin/selabel_lookup` behaves on RHEL8

So I believe we should be calling `Selinux.selabel_lookup(handle, file, 0)` and we can remove the code the stat calls, since we don't need the file mode.